pallets/pallet-bonded-coins/src/curves.rs

```diff
@@ -0,0 +1,253 @@
+		x.checked_mul(x).ok_or(ArithmeticError::Overflow)
+	}
+}
+
+#[derive(Clone, Debug, Encode, Decode, PartialEq, Eq, TypeInfo, MaxEncodedLen)]
+pub struct PolynomialFunctionParameters<F> {
+	pub m: F,
+	pub n: F,
+	pub o: F,
+}
+
+impl<F> BondingFunction<F> for PolynomialFunctionParameters<F>
+where
+	F: FixedSigned,
+{
+	fn calculate_costs(&self, low: F, high: F) -> Result<F, ArithmeticError> {
+		// Calculate high - low
+		let delta_x = high.checked_sub(low).ok_or(ArithmeticError::Underflow)?;
+
+		let high_low_mul = high.checked_mul(low).ok_or(ArithmeticError::Overflow)?;
+		let high_square = Self::square(high)?;
+		let low_square = Self::square(low)?;
+
+		// Factorized cubic term:  (high^2 + high * low + low^2)
+		let cubic_term = high_square
+			.checked_add(high_low_mul)
+			.ok_or(ArithmeticError::Overflow)?
+			.checked_add(low_square)
+			.ok_or(ArithmeticError::Overflow)?;
+
+		// Calculate m * (high^2 + high * low + low^2)
+		let term1 = self.m.checked_mul(cubic_term).ok_or(ArithmeticError::Overflow)?;
+
+		let high_plus_low = high.checked_add(low).ok_or(ArithmeticError::Overflow)?;
+
+		// Calculate n * (high + low)
+		let term2 = self.n.checked_mul(high_plus_low).ok_or(ArithmeticError::Overflow)?;
+
+		// Final calculation with factored (high - low)
+		let result = term1
+			.checked_add(term2)
+			.ok_or(ArithmeticError::Overflow)?
+			.checked_add(self.o)
+			.ok_or(ArithmeticError::Overflow)?;
+
+		result.checked_mul(delta_x).ok_or(ArithmeticError::Overflow)
+	}
+}
+
+#[derive(Clone, Debug, Encode, Decode, PartialEq, Eq, TypeInfo, MaxEncodedLen)]
+pub struct SquareRootFunctionParameters<F> {
+	pub m: F,
+	pub n: F,
+}
+
+impl<F> BondingFunction<F> for SquareRootFunctionParameters<F>
+where
+	F: FixedSigned + PartialOrd<I9F23> + From<I9F23> + ToFixed,
+{
+	fn calculate_costs(&self, low: F, high: F) -> Result<F, ArithmeticError> {
+		// Ensure that high and low are positive (logarithms of negative numbers are undefined)
+
+		// Calculate sqrt(high^3) and sqrt(low^3)
+		let sqrt_x3_high: F = sqrt::<F, F>(high)
+			.map_err(|_| ArithmeticError::Underflow)?
+			.checked_mul(high)
+			.ok_or(ArithmeticError::Overflow)?;
+
+		let sqrt_x3_low: F = sqrt::<F, F>(low)
+			.map_err(|_| ArithmeticError::Underflow)?
+			.checked_mul(low)
+			.ok_or(ArithmeticError::Overflow)?;
+
+		let delta_sqrt_x3 = sqrt_x3_high
+			.checked_sub(sqrt_x3_low)
+			.ok_or(ArithmeticError::Underflow)?;
+
+		let term1 = self.m.checked_mul(delta_sqrt_x3).ok_or(ArithmeticError::Overflow)?;
+
+		// Calculate n * (high - low) (linear term)
+		let delta_x = high.checked_sub(low).ok_or(ArithmeticError::Underflow)?;
+		let term2 = self.n.checked_mul(delta_x).ok_or(ArithmeticError::Overflow)?;
+
+		// Calculate the final result (sqrt + linear terms)
+		term1.checked_add(term2).ok_or(ArithmeticError::Overflow)
+	}
+}
+
+#[derive(Clone, Debug, Encode, Decode, PartialEq, Eq, TypeInfo, MaxEncodedLen)]
+pub struct LMSRFunctionParameters<F> {
```

Comment:
Especially here, you should move all lmsr-related stuff into its own module.

Reply:
[286e9dc](https://github.com/KILTprotocol/kilt-node/pull/765/commits/286e9dc057348e5a50004cbb6447aee0954b2754)